docs/runtime-user-guide.md

```diff
@@ -0,0 +1,119 @@
+# Runtime User Guide
+
+This document is focused on clarifying the expectations of the user
+container running in a Knative environment. This is separate from the expectations of the
+environment itself. The expectations of the environment are detailed in the [runtime
+contract](https://github.com/knative/serving/blob/master/docs/runtime-contract.md).
+
+The target audience of this document is _developers_ and _language and tooling
+developers_ as defined below:
+
+- **Developers** write code which is packaged into a container which is run on
+  the Knative cluster.
+- **Language and tooling developers** typically write tools used by
+    _developers_ to package code into containers. As such, they are concerned
+    that tooling which wraps developer code complies with this runtime contract.
+
+## General Behavior
+
+Containers running in Knative should be stateless containers. Stateless
+containers should have the following properties:
+
+- Fast startup time: The time from container launch to serving traffic should be
+  minmized to enable the platform to provide rapid autoscaling and scale-to-zero
+  semantics.
+- Minimize local state: The container may be killed and replaced to support
+  autoscaling and cluster maintenance operations. The performance or correctness
+  of the application should not rely on state persisted locally.
+- CPU usage only while requests are active: As containers may be killed or
+  suspended when no connections are active the container should only perform
+  operations that complete synchronously with the request's response.
+
+## Container Termination
+
+Knative containers should respond to `SIGTERM` signals to clean-up any resources
+necessary for a graceful shutdown including open streams.
+
+## Connection
+
+A Knative container must package a webserver that can respond to HTTP/1.1
+requests on a specified port.
+
+All traffic that comes to the container should be expected to pass through a
+proxy.
+
+### Ports
+
+All Knative containers must listen for incoming traffic on a specific port. A
+port may be specified through the API otherwise a platform-provided port will be chosen.
+
+The selected port will be made available within the container as the environment variable `$PORT`.
+It is recommended to consume the environment variable `$PORT` rather than hard-code a particular
+port within the application.
+
+Example (See
+[Container](https://github.com/knative/serving/blob/master/docs/spec/spec.md#container))
+```yaml
+...
+  containers:
+   - image: example-image:1.0
+     ports:
+      - containerPort: 8081
+...
+```
+
+### Protocols
+
+HTTP/1.1 is the default protocol for Knative. To use HTTP/2.0 you must specify
+`h2c` in the API using the port `name` field.
+
+Example (See
+[Container](https://github.com/knative/serving/blob/master/docs/spec/spec.md#container))
+```yaml
+...
+  containers:
+   - image: example-image:1.0
+     ports:
+      - name: "h2c"
+        containerPort: 8081
+...
+```
```

Comment:
Define here or in connection that the proxy terminates SSL and may convert HTTP versions automatically, so the user container only needs to speak a single HTTP version.

(I found it strange to have "Connection" and "Protocols" so far separated, FWIW.